spec.md

```diff
@@ -1096,6 +1010,30 @@ Ports are always considered from the perspective of "inside the module".
 Moreover, input ports may only appear "on the left side" of a connect, while output ports may only appear "on the right side" of a connect.
 Finally, while submodules instances and memories are strictly sources, they interact with the sub-field rule below, allowing connections to their input ports
 
+## Type Equivalence
+
+The type equivalence relation determines whether one values of one type can be coerced into values of another.
```

Comment:
I'm not sure if type coercion is the right concept here. Type equivalence is really this concept that is used narrowly to check the legality of statements or expressions. Put differently, the type equivalence relation says that `UInt<8>` is equal to `UInt`. This really only matters for connects to allow:

```
wire a: UInt
wire b: UInt<8>
connect b, a
```

However, the _type_ of that connect is still `(UInt<8>, UInt) -> ()`. I.e., the type of `a` is not coerced in any way. `a` is still an uninferred `UInt`.

Put differently again, this entire section is basically, "How do I write a verifier for connect ops?" as opposed to anything type theoretic.

It might be possible to reframe all this as "connects require exact type equivalence and uninferred is coereced to its inferred version". However, that is practically not what is happening.

Reply:
Maybe this entire section of type equivalence should really be moved under connects and called "connect legality" or "connect verification". 🤔

Reply:
I had originally considered rewriting this section, but it will need some dedicated attention.

I'll change this intro back to the original copy.